File: outputs/src/@opencode/sdk/src/session.ts
/**
 * Session management for OpenCode SDK.
 *
 * Provides types and operations for managing OpenCode sessions.
 */

import type { MessageRole, SessionState } from './types';
export type { SessionState } from './types';

/**
 * Session information returned by the API.
 */
export interface SessionInfo {
  /** Unique session identifier. */
  id: string;
  /** Session creation timestamp (ISO 8601). */
  createdAt: string;
  /** Last update timestamp (ISO 8601). */
  updatedAt: string;
  /** Number of messages in the session. */
  messageCount: number;
  /** Preview of the last message (if available). */
  preview?: string;
}

/**
 * Session creation request.
 */
export interface CreateSessionRequest {
  /** Optional initial prompt to start the session with. */
  initialPrompt?: string;
}

/**
 * Session creation response from the API.
 */
export interface CreateSessionResponse {
  /** The created session ID. */
  sessionId: string;
  /** Creation timestamp. */
  createdAt: string;
  /** Session status. */
  status: string;
  /** Number of messages. */
  messageCount: number;
}

/**
 * Fork session request.
 */
export interface ForkSessionRequest {
  /** Message index to fork at. */
  forkAtMessageIndex: number;
}

/**
 * Fork session response.
 */
export interface ForkSessionResponse {
  /** New session ID after fork. */
  id: string;
  /** Parent session ID. */
  parentSessionId?: string;
  /** Message count in forked session. */
  messageCount: number;
}

/**
 * Add message request.
 */
export interface AddMessageRequest {
  /** Message role (user, assistant, system). */
  role?: string;
  /** Message content. */
  content: string;
}

/**
 * Add message response.
 */
export interface AddMessageResponse {
  /** Session ID. */
  sessionId: string;
  /** Total message count after adding. */
  messageCount: number;
}

/**
 * Session message type.
 */
export interface SessionMessage {
  /** Message role. */
  role: MessageRole;
  /** Message content. */
  content: string;
}

/**
 * Session summary information.
 */
export interface SessionSummary {
  /** Summary text. */
  summary: string;
  /** Summary creation timestamp. */
  createdAt: string;
}

/**
 * SDK Session representation for local use.
 */
export class SdkSession {
  /** Unique session identifier. */
  public id: string;
  /** Session messages. */
  public messages: SessionMessage[];
  /** Creation timestamp. */
  public createdAt: Date;
  /** Last update timestamp. */
  public updatedAt: Date;
  /** Current session state. */
  public state: SessionState;

  /**
   * Creates a new SDK session with the given ID.
   */
  constructor(id: string) {
    this.id = id;
    this.messages = [];
    this.createdAt = new Date();
    this.updatedAt = new Date();
    this.state = 'idle';
  }

  /**
   * Adds a user message to the session.
   */
  addUserMessage(content: string): void {
    this.messages.push({ role: 'user', content });
    this.updatedAt = new Date();
  }

  /**
   * Adds an assistant message to the session.
   */
  addAssistantMessage(content: string): void {
    this.messages.push({ role: 'assistant', content });
    this.updatedAt = new Date();
  }

  /**
   * Adds a system message to the session.
   */
  addSystemMessage(content: string): void {
    this.messages.push({ role: 'system', content });
    this.updatedAt = new Date();
  }

  /**
   * Returns the number of messages.
   */
  get messageCount(): number {
    return this.messages.length;
  }
}

/**
 * Parse role string to MessageRole.
 */
export function parseRole(role?: string): MessageRole {
  switch (role?.toLowerCase()) {
    case 'assistant':
      return 'assistant';
    case 'system':
      return 'system';
    default:
      return 'user';
  }
}

File: rust-opencode-port/crates/sdk/src/session.rs
//! Session management for OpenCode SDK.
//!
//! Provides types and operations for managing OpenCode sessions.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::error::SdkError;

/// Session information returned by the API.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionInfo {
    /// Unique session identifier.
    pub id: Uuid,

    /// Session creation timestamp.
    pub created_at: DateTime<Utc>,

    /// Last update timestamp.
    pub updated_at: DateTime<Utc>,

    /// Number of messages in the session.
    #[serde(default)]
    pub message_count: usize,

    /// Preview of the last message (if available).
    #[serde(default)]
    pub preview: Option<String>,
}

/// Session creation request.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateSessionRequest {
    /// Optional initial prompt to start the session with.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub initial_prompt: Option<String>,
}

/// Session creation response from the API.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateSessionResponse {
    /// The created session ID.
    pub session_id: Uuid,

    /// Creation timestamp.
    pub created_at: String,

    /// Session status.
    pub status: String,

    /// Number of messages.
    pub message_count: usize,
}

/// Fork session request.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ForkSessionRequest {
    /// Message index to fork at.
    pub fork_at_message_index: usize,
}

/// Fork session response.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ForkSessionResponse {
    /// New session ID after fork.
    pub id: Uuid,

    /// Parent session ID.
    #[serde(default)]
    pub parent_session_id: Option<String>,

    /// Message count in forked session.
    pub message_count: usize,
}

/// Add message request.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AddMessageRequest {
    /// Message role (user, assistant, system).
    #[serde(default = "default_role")]
    pub role: String,

    /// Message content.
    pub content: String,
}

fn default_role() -> String {
    "user".to_string()
}

/// Add message response.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AddMessageResponse {
    /// Session ID.
    pub session_id: Uuid,

    /// Total message count after adding.
    pub message_count: usize,
}

/// Session summary information.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionSummary {
    /// Summary text.
    pub summary: String,

    /// Summary creation timestamp.
    pub created_at: DateTime<Utc>,
}

/// Message type for session messages.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionMessage {
    /// Message role.
    pub role: String,

    /// Message content.
    pub content: String,
}

impl SessionInfo {
    /// Creates a new SessionInfo with the given ID.
    pub fn new(id: Uuid) -> Self {
        let now = Utc::now();
        Self {
            id,
            created_at: now,
            updated_at: now,
            message_count: 0,
            preview: None,
        }
    }

    /// Validates that the session ID is valid.
    pub fn validate_id(id: &str) -> Result<Uuid, SdkError> {
        Uuid::parse_str(id).map_err(|_| SdkError::session_not_found(id))
    }
}

impl SdkSession {
    /// Creates a new SDK session with the given ID.
    pub fn new(id: Uuid) -> Self {
        Self {
            id,
            messages: Vec::new(),
            created_at: Utc::now(),
            updated_at: Utc::now(),
            state: SessionState::Idle,
        }
    }

    /// Adds a user message to the session.
    pub fn add_user_message(&mut self, content: impl Into<String>) {
        self.messages.push(SessionMessage {
            role: "user".to_string(),
            content: content.into(),
        });
        self.updated_at = Utc::now();
    }

    /// Adds an assistant message to the session.
    pub fn add_assistant_message(&mut self, content: impl Into<String>) {
        self.messages.push(SessionMessage {
            role: "assistant".to_string(),
            content: content.into(),
        });
        self.updated_at = Utc::now();
    }
}

/// SDK Session representation for local use.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SdkSession {
    /// Unique session identifier.
    pub id: Uuid,

    /// Session messages.
    pub messages: Vec<SessionMessage>,

    /// Creation timestamp.
    pub created_at: DateTime<Utc>,

    /// Last update timestamp.
    pub updated_at: DateTime<Utc>,

    /// Current session state.
    pub state: SessionState,
}

/// Session state enumeration.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SessionState {
    /// Session is idle and ready for input.
    Idle,

    /// Session is processing a request.
    Processing,

    /// Session has been aborted.
    Aborted,

    /// Session has completed.
    Completed,

    /// Session encountered an error.
    Error,
}

impl Default for SessionState {
    fn default() -> Self {
        Self::Idle
    }
}

impl std::fmt::Display for SessionState {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Idle => write!(f, "idle"),
            Self::Processing => write!(f, "processing"),
            Self::Aborted => write!(f, "aborted"),
            Self::Completed => write!(f, "completed"),
            Self::Error => write!(f, "error"),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_session_info_new() {
        let id = Uuid::new_v4();
        let info = SessionInfo::new(id);
        assert_eq!(info.id, id);
        assert_eq!(info.message_count, 0);
        assert!(info.preview.is_none());
    }

    #[test]
    fn test_session_info_validate_id() {
        let id = Uuid::new_v4();
        let result = SessionInfo::validate_id(&id.to_string());
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), id);

        let result = SessionInfo::validate_id("invalid");
        assert!(result.is_err());
    }

    #[test]
    fn test_sdk_session_new() {
        let id = Uuid::new_v4();
        let session = SdkSession::new(id);
        assert_eq!(session.id, id);
        assert!(session.messages.is_empty());
        assert_eq!(session.state, SessionState::Idle);
    }

    #[test]
    fn test_sdk_session_add_messages() {
        let mut session = SdkSession::new(Uuid::new_v4());
        session.add_user_message("Hello");
        session.add_assistant_message("Hi there!");

        assert_eq!(session.messages.len(), 2);
        assert_eq!(session.messages[0].role, "user");
        assert_eq!(session.messages[1].role, "assistant");
    }

    #[test]
    fn test_session_state_display() {
        assert_eq!(SessionState::Idle.to_string(), "idle");
        assert_eq!(SessionState::Processing.to_string(), "processing");
        assert_eq!(SessionState::Aborted.to_string(), "aborted");
    }
}
